azerothcore/azerothcore-wotlk
# Fix Scholomance Occultists behavior

## Changes Proposed:
- Fixed an issue that was causing UpdateEntry mobs to not be lootable.
- Added script for handling Scholomance Occultist behavior.
  - In human form use Bone Armor ever 60 sec
  - In human form use Counterspell on targets in threatlist that are casting.
  - In human form use Drain Mana on random caster target.
  - Update Entry when below 30% hp to Dark Shade.
  - Use Shadowbolt volley spell cast only in Dark Shade form.
  - On reset in Dark Shade form acquire normal Occultist form.

### More about the UpdateEntry loot related bug:
When UpdateEntry is called it sets the hp of the updated entry to the previous health that the creature had.
When this occurs it sets m_PlayerDamageReq to a much higher value than it should be and in terms makes it so Unit::Kill isRewardAllowed to be false and thus preventing the player to loot the creature.

## Issues Addressed:
https://github.com/azerothcore/azerothcore-wotlk/issues/4520
- Closes 3. Occultists don't turn into shades


## SOURCE:

https://youtu.be/s-2GXud7CJI?t=1496

## Tests Performed:

- Builds without errors.
- Tested in game on linux/docker setup from windows client.

## How to Test the Changes:

##### How to reproduce the UpdateEntry loot bug:
1. Add SAI to a creature in db with Update Template action when mob is below 30%. (Make sure you backup the creature query so you can revert)
2. Note it should be a creature of the same level and stats.
3. Kill the creature and check if you can loot the corpse. (You shouldn't be able to)
4. Apply the fix from [commit](https://github.com/azerothcore/azerothcore-wotlk/pull/5314/commits/dd4e6f542b555fac79718272788642bdaeffea33) and repeat the above 3 steps.
5. Check if you can now loot the corpse. (You should be able to)

Before the changes:
![2021-04-16 16-10-11](https://user-images.githubusercontent.com/10816895/115039130-d8103f00-9ed8-11eb-92fd-3e65a274f518.gif)

After the changes:
![2021-04-16 16-14-04](https://user-images.githubusercontent.com/10816895/115039256-f5dda400-9ed8-11eb-8676-06195dba112b.gif)

##### Testing Scholomance Occultists behavior adjustments:
1. Switch to this branch and rebuilt.
2. Run the query from [commit](https://github.com/azerothcore/azerothcore-wotlk/pull/5314/commits/e402021166c5eeca247557a1bd489d5d0dc089ac)
3. Log in and go to Scholomance - The Coven
4. Attack a Scholomance Occultist.
5. Check if he is casting Bone Armor spell.
6. Check if he is casting Drain Mana.
7. Check if he will interrupt your spell casts if he isn't casting.
8. Check if he transforms to Dark Shade when the Occultist is below 30%.
9. Check if he is immune to physical in Dark Shade form.
10. Check if the mob name updates to Dark Shade.
11. Check if when reset/evade he returns to his normal human form.
12. Check if when reset/evade the mob updates his name to Scholomance Occultist.
13. Check if when killed in Dark Shade he is lootable.
![2021-04-16 12-48-14](https://user-images.githubusercontent.com/10816895/115039719-5ff64900-9ed9-11eb-927e-00b31415b7a1.gif)
![2021-04-16 12-47-32](https://user-images.githubusercontent.com/10816895/115039737-64bafd00-9ed9-11eb-9e4e-3c0685d38bfa.gif)

## Known Issues and TODO List:



## Target Branch(es):







 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
